model/src/payouts/quanto.rs

```diff
@@ -0,0 +1,609 @@
+    ShortOwesTooMuch { owing: Amount, margin: Amount },
+    #[error("{0}")]
+    Other(#[from] anyhow::Error),
+}
+
+impl Curve {
+    pub fn new(
+        initial_price: u64,
+        n_contracts: u64,
+        leverage_long: Leverage,
+        leverage_short: Leverage,
+        n_payouts: usize,
+        multiplier: Decimal,
+        fee_offset: CompleteFee,
+    ) -> Self {
+        Self {
+            n_payouts,
+            initial_price,
+            n_contracts,
+            leverage_long,
+            leverage_short,
+            multiplier,
+            fee_offset,
+        }
+    }
+
+    /// Discretize the payout curve into distinct payouts.
+    pub fn discretized_payouts(&self) -> Result<Vec<Payout>, Error> {
+        let initial_margin_long = self.initial_margin_long()?;
+        let initial_margin_short = self.initial_margin_short()?;
+        let initial_margin_total = initial_margin_long + initial_margin_short;
+
+        match self.fee_offset {
+            CompleteFee::LongPaysShort(long_owes) if long_owes > initial_margin_long => {
+                return Err(Error::LongOwesTooMuch {
+                    owing: long_owes,
+                    margin: initial_margin_long,
+                });
+            }
+            CompleteFee::ShortPaysLong(short_owes) if short_owes > initial_margin_short => {
+                return Err(Error::ShortOwesTooMuch {
+                    owing: short_owes,
+                    margin: initial_margin_short,
+                });
+            }
+            _ => {}
+        }
+
+        let long_liquidation_interval = self
+            .long_liquidation_interval()
+            .context("Could not calculate long liquidation interval")?;
+        let short_liquidation_interval = self
+            .short_liquidation_interval()
+            .context("Could not calculate short liquidation interval")?;
+
+        let long_liquidation_threshold = long_liquidation_interval.end();
+        let short_liquidation_threshold = short_liquidation_interval.start();
+
+        let long_liquidation_payout = Payout {
+            interval: long_liquidation_interval.clone(),
+            long: Amount::ZERO,
+            short: initial_margin_total,
+        };
+
+        let mut payouts = vec![long_liquidation_payout];
+
+        // We want to split the settlement region of the curve into `n-payouts - 2` segments
+        let step = Decimal::from(short_liquidation_threshold - long_liquidation_threshold)
+            / Decimal::from(self.n_payouts - 2);
+        let step = step.to_u64().context("Could not convert step to u64")?;
+
+        // We start building the settlement intervals right after the end of the long liquidation
+        // interval
+        let mut prev_upper_bound = *long_liquidation_threshold;
+        loop {
+            let lower_bound = prev_upper_bound + 1;
+            let upper_bound = lower_bound + step;
```

Comment:
The base point to start computing what I've called the "settlement payouts" is the end of the long liquidation interval. Because our payout price intervals are _inclusive_, I have to set the first lower bound of these to `long_liquidation_threshold + 1` to avoid an overlap.

> This would mean the interval will almost always result in the "special case" that the last interval is smaller!

I don't see how you got to this conclusion. Can you explain your reasoning?

> I would have expected that we say
>
> let upper_bound = lower_bound + step - 1; in the loop

What I see here is that you propose to make the `step` smaller by 1, but I don't understand why.

Keep in mind that when we calculate the `step` we already take into account the fact that we are starting from `long_liquidation_threshold + 1`:

```
step = (short_liquidation_threshold - long_liquidation_threshold) / (n_payouts - 2)
     = ((short_liquidation_threshold + 1) - (long_liquidation_threshold - 1 ) / (n_payouts - 2) // accounting for inclusive intervals
```

> I created [gsheet](https://docs.google.com/spreadsheets/d/1A59TL0rcFmBet_y-2N5s9tXngf_Ifrx652pcgCy09Rk/edit?usp=sharing) to simulate the behavior.

I can't quite parse what's happening in that one :(

Btw, with your suggestion the last settlement interval is still smaller than the others _and_ we produce an extra CET. It is inevitable that we will more often than not have a smaller settlement interval, because the overall settlement interval will usually not be divisible by the step.

Reply:
> The base point to start computing what I've called the "settlement payouts" is the end of the long liquidation interval. Because our payout price intervals are _inclusive_, I have to set the first lower bound of these to `long_liquidation_threshold + 1` to avoid an overlap.
> 
> > This would mean the interval will almost always result in the "special case" that the last interval is smaller!
> 
> I don't see how you got to this conclusion. Can you explain your reasoning?

I think the comment stating that it is a special case 

```
        // We have to consider a special case if the last settlement interval is smaller than every
        // other settlement interval
```

lead me to believe that this would not be something that can happen.

< [snip]
> Btw, with your suggestion the last settlement interval is still smaller than the others _and_ we produce an extra CET (I tested it using the snapshot test). It is inevitable that we will more often than not have a smaller settlement interval, because the overall settlement interval will usually not be divisible by the step.

True, my suggestion is incorrect, because it is based on the assumption of `n_payouts + 1` then 🙈
Yes, you are right, there is likely always some remainder, and that is just one of the intervals.
Sorry, I hope what I said was not too confusing.

Question on top, is it "fair" that the shorter interval is the last interval before the `short` liquidation? Should that interval not be the one closest to the original price (i.e. in the middle of the distribution) to be fair? (not super important)

Reply:
> Question on top, is it "fair" that the shorter interval is the last interval before the short liquidation? Should that interval not be the one closest to the original price (i.e. in the middle of the distribution) to be fair? (not super important)

It's important to highlight that no matter where the short settlement interval lies, the liquidation intervals remain intact. So in that sense it makes no difference.

Any slightly shorter settlement interval just means that a particular region has a bit more resolution in terms of deciding the payouts. The fairest approach would be to randomise the location of the shorter settlement interval, but the complexity cost far outweighs the benefits in my opinion.